pkg/download/download.go

```diff
@@ -0,0 +1,243 @@
+package download
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/credentials"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/firmwareci/contest-client/pkg/env"
+	"github.com/jlaffaye/ftp"
+	"github.com/pkg/sftp"
+	"golang.org/x/crypto/ssh"
+)
+
+// DownloadBinary parses the unparsed URL, and proceeds to download
+func DownloadBinary(unparsedURL string) (string, error) {
+	URL, err := url.Parse(unparsedURL)
+	if err != nil {
+		return "", err
+	}
+	filename := path.Base(URL.Path)
+
+	binaryDirectory := os.Getenv(env.EnvBinDir)
+
+	if binaryDirectory != "" {
+		if path.IsAbs(binaryDirectory) != true {
+			return "", fmt.Errorf("Path: '%s', is not absolute", binaryDirectory)
+		}
+	} else {
+		binaryDirectory = env.DefaultBinaryDir
+	}
+
+	binaryPath := path.Join(binaryDirectory, filename)
+
+	file, err := os.Create(binaryPath)
+	if err != nil {
+		return "", err
+	}
+
+	defer file.Close()
+
+	switch URL.Scheme {
+	case "http", "https", "":
+		{
+			if err := downloadHTTP(file, URL); err != nil {
+				return "", err
+			}
+		}
+	case "ftp":
+		{
+			if err := downloadFTP(file, URL); err != nil {
+				return "", err
+			}
+		}
+	case "sftp":
+		{
+			if err := downloadSFTP(file, URL); err != nil {
+				return "", err
+			}
+		}
+	case "S3":
+		{
+			if err := downloadS3(file, URL); err != nil {
+				return "", err
+			}
+		}
+
+	}
+	return binaryPath, nil
+}
+
+// DownloadHttp downloads the file at the end of the Url via http/ https,
+// into the given file. Both pointers owned by caller.
+func downloadHTTP(file *os.File, URL *url.URL) error {
+	resp, err := http.Get(URL.String())
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if _, err := io.Copy(file, resp.Body); err != nil {
+		return err
+	}
+	return nil
+}
+
+// DownloadFtp downloads the file at the end of the Url via ftp,
+// into the given file. Both pointers owned by caller.
+func downloadFTP(file *os.File, URL *url.URL) error {
+	if URL.User == nil {
+		URL.User = url.UserPassword("anonymous", "anonymous")
+	}
+
+	connection, err := ftp.DialTimeout(URL.Hostname(), 60*time.Second)
+	if err != nil {
+		return err
+	}
+
+	defer connection.Quit()
+
+	pw, set := URL.User.Password()
```

Comment:
It's just a nit: I would check User and Password first, and then connect. But that's personal preference here.

Reply:
Sure